kubespawner/proxy.py

```diff
@@ -0,0 +1,244 @@
+import os
+import string
+import escapism
+import json
+from kubernetes import client
+
+from jupyterhub.proxy import Proxy
+from jupyterhub.utils import exponential_backoff
+
+from kubespawner.objects import make_ingress
+from kubespawner.utils import generate_hashed_slug
+from kubespawner.reflector import NamespacedResourceReflector
+from concurrent.futures import ThreadPoolExecutor
+from traitlets import Unicode
+from tornado import gen
+from tornado.concurrent import run_on_executor
+
+
+class IngressReflector(NamespacedResourceReflector):
+    kind = 'ingresses'
+    labels = {
+        'heritage': 'jupyterhub',
+        'component': 'singleuser-server',
+        'hub.jupyter.org/proxy-route': 'true'
+    }
+
+    list_method_name = 'list_namespaced_ingress'
+    api_group_name = 'ExtensionsV1beta1Api'
+
+    @property
+    def ingresses(self):
+        return self.resources
+
+class ServiceReflector(NamespacedResourceReflector):
+    kind = 'services'
+    labels = {
+        'heritage': 'jupyterhub',
+        'component': 'singleuser-server',
+        'hub.jupyter.org/proxy-route': 'true'
+    }
+
+    list_method_name = 'list_namespaced_service'
+
+    @property
+    def services(self):
+        return self.resources
+
+class EndpointsReflector(NamespacedResourceReflector):
+    kind = 'endpoints'
+    labels = {
+        'heritage': 'jupyterhub',
+        'component': 'singleuser-server',
+        'hub.jupyter.org/proxy-route': 'true'
+    }
+
+    list_method_name = 'list_namespaced_endpoints'
+
+    @property
+    def endpoints(self):
+        return self.resources
+
+class KubeIngressProxy(Proxy):
+    namespace = Unicode(
+        config=True,
+        help="""
+        Kubernetes namespace to spawn ingresses for single-user servers in.
```

Comment:
Perhaps: `Kubernetes namespace, specified by ingress, to spawn single-user servers in.`

Reply:
Each single-user server has an ingress associated with it, and both of these things belong to a namespace. This is just for creating ingress objects that corrsepond to spawned servers. I'm confused by the alternate suggestion and am unsure what it means :)

Reply:
Is this accurate? What's confusing me is I thought ingresses are essentially rules.

Kubernetes namespace for both single users servers to be spawned and their ingresses.

Reply:
Everything in kubernetes is an object that represents something. In this case, Ingresses are objects that represent HTTP routing rules.

This only sets the namespace for the ingresses to be created, though. The namespace for singleuser servers is specified in KubeSpawner.namespace, rather than KubeProxy.namespace.

Reply:
How about:

"Kubernetes namespace to create Ingress objects in.

This corresponds to KubeSpawner.namespace, which sets the namespace single-user servers
are spawned in. For most use cases, these two should be the same
"

Reply:
Love the suggestion. Thanks for the explanation @yuvipanda.